### ISSUE TITLE
System.Text.Json.JsonReader is not reading numbers correctly

### ISSUE THREAD

**Issue by Gonkers** [dotnet/runtime#54096]
### Description
System.Text.Json cannot read JSON that just consists of just a number. According to the [specifications](https://www.json.org/json-en.html), this should be valid.

```
System.Text.Json.JsonReaderException
  HResult=0x80131500
  Message='0x00' is an invalid end of a number. Expected 'E' or 'e'. LineNumber: 0 | BytePositionInLine: 5.
  Source=System.Text.Json
```

Here are some unit tests to reproduce the issue.

```c#
        public class NumberValueTests
        {
            public static JsonTokenType GetTokenType(string jsonValueOnly, byte[] buffer)
            {
                var bytesEncoded = Encoding.UTF8.GetBytes(jsonValueOnly, buffer);
                var jsonReaderOptions = new JsonReaderOptions { AllowTrailingCommas = true, CommentHandling = JsonCommentHandling.Skip };
                var readerState = new JsonReaderState(jsonReaderOptions);
                var reader = new Utf8JsonReader(buffer, false, readerState);
                reader.Read();

                return reader.TokenType;
            }


            [Fact]
            public void  NumberValues()
            {
                Assert.Equal(JsonTokenType.String, GetTokenType("\"4444\"", new byte[10]));
                Assert.Equal(JsonTokenType.False, GetTokenType("false", new byte[10]));
                Assert.Equal(JsonTokenType.True, GetTokenType("true", new byte[10]));
                Assert.Equal(JsonTokenType.Null, GetTokenType("null", new byte[10]));

                // !!!  THESE SHOULD BE NUMBER  !!!
                Assert.Equal(JsonTokenType.Number, GetTokenType("44.44", new byte[5]));
                Assert.Equal(JsonTokenType.Number, GetTokenType("4444", new byte[4]));

                // !!!  THIS SHOULD NOT THOW AN INTERNAL EXCEPTION  !!!
                Assert.Throws<System.Text.Json.JsonReaderException>(() => GetTokenType("4444", new byte[10]));
            }
        }
```

### Configuration
* Which version of .NET is the code running on? `5.0.300`
* What OS and version, and what distro if applicable? `Windows 10`
* What is the architecture (x64, x86, ARM, ARM64)? `x64`
* Do you know whether it is specific to that configuration? `I don't know.`

**Comment by Tornhoof** [comment]
You're ignoring the length of ``GetBytes``, so you're passing the ``Utf8JsonReader`` the full byte array and not the sliced one. You need to slice your buffer properly.
``var reader = new Utf8JsonReader(buffer.AsSpan(0, bytesEncoded), false, readerState);``
That it happens to work for the other types is because string has a delimiter symbol and null, false/true a specific fixed length tokens.
See https://github.com/dotnet/runtime/issues/29947 for someone with the same problem.

As for specification reference, see https://datatracker.ietf.org/doc/html/rfc8259#section-2, whitespace is defined as:
```ini
ws = *(
              %x20 /              ; Space
              %x09 /              ; Horizontal tab
              %x0A /              ; Line feed or New line
              %x0D )              ; Carriage return
```
and this excludes 0x00 bytes.

**Comment by Gonkers** [comment]
@Tornhoof, you seem to have misunderstood my tests. Let me explain further.

1)
   ```c#
   Assert.Equal(JsonTokenType.Number, GetTokenType("44.44", new byte[5]));
   Assert.Equal(JsonTokenType.Number, GetTokenType("4444", new byte[4]));
   ```
   These lines produce `JsonTokenType.None` _**not**_ `JsonTokenType.Number` like they should. These tests fail. _The `byte[]` are the exact length length they need to be._ Which you claim that I am not doing.
   
1)  
   ```
   Assert.Throws<System.Text.Json.JsonReaderException>(() => GetTokenType("4444", new byte[10]));
   ```
   This line does not compile, it generates an error. The exception thrown here by the call `GetTokenType("4444", new byte[10])`  is from an [internal class named `JsonReaderException`](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/src/System/Text/Json/Reader/JsonReaderException.cs#L10).  I cannot use a `try...catch` block to catch the exception specifically because `JsonReaderException` does not exist to me. This is a bug. 
   ```
   The type or namespace name 'JsonReaderException' does not exist in the namespace 'System.Text.Json' (are you missing an assembly reference?)
   ```
   In summary, this unit test _should not pass_
   ```c#
   try
   {
       GetTokenType("4444", new byte[10]);
   }
   catch (System.Exception ex)
   {
       Assert.Equal("System.Text.Json.JsonReaderException", ex.GetType().FullName);
   }
   ```

I provided unit tests to demonstrate the bugs. Please in the future run them to give us users the benefit of the doubt.

Thanks.

**Comment by Gonkers** [comment]
So to catch this `internal` exception I guess I'll need to do this...
`catch (JsonException ex) when (ex.GetType().FullName.Equals("System.Text.Json.JsonReaderException"))`

And I'll need to add spaces to all of my JSON so `JsonReader` doesn't throw exceptions...
```c#
Assert.Equal(JsonTokenType.Number, GetTokenType("44.44" + " ", new byte[6]));
Assert.Equal(JsonTokenType.Number, GetTokenType("4444" + " ", new byte[5]));
```

**Comment by davidfowl** [comment]
```
var bytesEncoded = Encoding.UTF8.GetBytes(jsonValueOnly, buffer);
```

This valid isn't used in `GetTokenType`, that's what @Tornhoof was getting at.

**Comment by Tornhoof** [comment]
>> I provided unit tests to demonstrate the bugs. Please in the future run them to give us users the benefit of the doubt.

Actually I did, that's how I saw your problem with the slice. The other problem your code has, is that you set finalblock to false ``new Utf8JsonReader(buffer, false, readerState);``. According to the comment for that parameter it specifies that no other data is following it, I admit I forgot that in my comment, as I just added the slice and not the true.
If you pass true and fix the buffer slice the tests pass.
https://github.com/dotnet/runtime/blob/4aa29f43fd15e0217ecaae5a06d1e355d0471bb0/src/libraries/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs#L197-L199

As for the exception, ``JsonReaderException`` is derived from ``JsonException`` that's the one you want to catch.

**Comment by Gonkers** [comment]
Okay please humor me. I do really appreciate you all responding. Can you please tell me if any of these json files are invalid? https://github.com/Gonkers/HowDoesJsonWork

I can read all of these files. I CANNOT read the `number.json` file without throwing an exception.
```c#
[Fact]
public void ReadingFiles()
{
    var dir = new DirectoryInfo(@"C:\Users\Gonke\Dev\HowDoesJsonWork");
    var jsonFiles = dir.GetFiles("*.json");
    foreach (var file in jsonFiles)
    {
        using var jsonStream = file.Open(FileMode.Open, FileAccess.Read, FileShare.Read);
        var buffer = new byte[4096];
        var bytesRead = jsonStream.Read(buffer);
        var jsonReaderOptions = new JsonReaderOptions { AllowTrailingCommas = true, CommentHandling = JsonCommentHandling.Skip };
        var readerState = new JsonReaderState(jsonReaderOptions);
        var reader = new Utf8JsonReader(buffer, false, readerState);
        var readSucessful = reader.Read();
        Assert.True(readSucessful);
    }
}
```

**Comment by Tornhoof** [comment]
Your files are all valid, your code is just not valid, as described before your code is ignoring the length of the file (in your examples) and you're passing ``false`` into the Utf8JsonReader.
Try
```csharp
[Fact]
public void ReadingFiles()
{
	var dir = new DirectoryInfo(@"D:\Repos\HowDoesJsonWork");
	var jsonFiles = dir.GetFiles("*.json");
	foreach (var file in jsonFiles)
	{
		using var jsonStream = file.Open(FileMode.Open, FileAccess.Read, FileShare.Read);
		var buffer = new byte[4096];
		var bytesRead = jsonStream.Read(buffer);
		var jsonReaderOptions = new JsonReaderOptions { AllowTrailingCommas = true, CommentHandling = JsonCommentHandling.Skip };
		var readerState = new JsonReaderState(jsonReaderOptions);
		var reader = new Utf8JsonReader(buffer.AsSpan(0, bytesRead), true, readerState);
		var readSucessful = reader.Read();
		Assert.True(readSucessful);
	}
}
```
As for why your bug happens only for number, as I said, your other examples are either fixed length tokens (null, false,true) or have boundary characters (string)

**Comment by Gonkers** [comment]
> > I provided unit tests to demonstrate the bugs. Please in the future run them to give us users the benefit of the doubt.
> 
> Actually I did, that's how I saw your problem with the slice. The other problem your code has, is that you set finalblock to false `new Utf8JsonReader(buffer, false, readerState);`. According to the comment for that parameter it specifies that no other data is following it, I admit I forgot that in my comment, as I just added the slice and not the true.
> If you pass true and fix the buffer slice the tests pass.
> https://github.com/dotnet/runtime/blob/4aa29f43fd15e0217ecaae5a06d1e355d0471bb0/src/libraries/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.cs#L197-L199
> 
> As for the exception, `JsonReaderException` is derived from `JsonException` that's the one you want to catch.

OMG Okay I get it now. While I was coding up that last example I didn't even see your reply come in. I get it now. Sorry for the confusion. Thanks.

**Comment by Gonkers** [comment]
This is a note for the next person to come along. This is a more "correct" example.  When you're reading your stream you need to detect if you're at the end of if it or if there is possibly more to come. Make note of the following line in the example.
```
var isFinalBlock = bytesRead < buffer.Length;
```
Also, if `reader.Read()` ever returns false, it means you've likely reached the end of your buffer and you need to get more data. (not shown in example).
```c#
[Fact]
public void ReadingFiles()
{
    var dir = new DirectoryInfo(@"C:\Users\Gonke\Dev\HowDoesJsonWork");
    var jsonFiles = dir.GetFiles("*.json");
    foreach (var file in jsonFiles)
    {
        using var jsonStream = file.Open(FileMode.Open, FileAccess.Read, FileShare.Read);
        var buffer = new byte[4096];
        var bytesRead = jsonStream.Read(buffer);
        var jsonReaderOptions = new JsonReaderOptions { AllowTrailingCommas = true, CommentHandling = JsonCommentHandling.Skip };
        var readerState = new JsonReaderState(jsonReaderOptions);
        var isFinalBlock = bytesRead < buffer.Length;
        var reader = new Utf8JsonReader(buffer.AsSpan(0, bytesRead), isFinalBlock, readerState);
        var readSucessful = reader.Read();

        Assert.True(readSucessful);
    }
}
```
I cannot add files with .json extension so here are the samples.
[false.txt](https://github.com/dotnet/runtime/files/6642738/false.txt)
[null.txt](https://github.com/dotnet/runtime/files/6642739/null.txt)
[number.txt](https://github.com/dotnet/runtime/files/6642740/number.txt)
[string.txt](https://github.com/dotnet/runtime/files/6642741/string.txt)
[true.txt](https://github.com/dotnet/runtime/files/6642742/true.txt)

**Comment by Gonkers** [comment]
An additional note for future readers, I was adapting [this example](https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-use-dom-utf8jsonreader-utf8jsonwriter#read-from-a-stream-using-utf8jsonreader) and it uses a static `false` for the `isFinalBlock` parameter. I am filing a bug with the docs team to see if the example can be updated.
